devito/operator.py

```diff
@@ -109,32 +108,81 @@ def __init__(self, expressions, **kwargs):
         self.dle_flags = dle_state.flags
         self.func_table.update(OrderedDict([(i.name, MetaCall(i, True))
                                             for i in dle_state.elemental_functions]))
-        parameters.extend([i.argument for i in self.dle_arguments])
         self.dimensions.extend([i.argument for i in self.dle_arguments
                                 if isinstance(i.argument, Dimension)])
         self._includes.extend(list(dle_state.includes))
 
         # Introduce the required symbol declarations
         nodes = iet_insert_C_decls(dle_state.nodes, self.func_table)
 
-        # Initialise ArgumentEngine
-        self.argument_engine = ArgumentEngine(clusters.ispace, parameters,
-                                              self.dle_arguments)
+        # Insert data and pointer casts for array parameters and profiling structs
+        casts = [ArrayCast(f) for f in self.input if f.is_Tensor and f._mem_external]
+        profiler = Object(self.profiler.name, self.profiler.dtype, self.profiler.new)
+        casts.append(PointerCast(profiler))
+        nodes = (List(body=casts), nodes)
+
+        # Derive parameters as symbols not defined in the kernel itself
+        parameters = derive_parameters(nodes)
 
-        parameters = self.argument_engine.arguments
+        # Filter all internally-allocated temporary `Array` types
+        # TODO: Huge hack, but to fix it, allocations need to be part of IET
```

Comment:
OK, I've thought about this some more. Perhaps it would be better to let ``derive_parameters`` take an optional parameter that, if set to True, drops all local arrays? It would kinda make sense to me